### ISSUE TITLE
Bootsrap fails during replace operation while starting "off-strategy compaction". Huge amount of "Error applying view update" errors were received

### ISSUE THREAD

**Issue by juliayakovlev** [scylladb/scylladb#12693]
## Issue description

- [x] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

This issue happened during terminate and replace nemesis.
Node `longevity-tls-50gb-3d-master-db-node-91f4614d-2 (10.12.8.179)` has been terminated. Replacing with node `longevity-tls-50gb-3d-master-db-node-91f4614d-7` was started.

`off-strategy compaction` was started for a view (mview.users_by_last_name) during bootstrap and found 643 candidates.
Then syslog-ng informed that "queue full, dropping message" and syslog-ng got error: "I/O error occurred error='Broken pipe":
```
t:2023-01-28 07:53:36,916 f:db_log_reader.py l:113 c:sdcm.db_log_reader p:DEBUG &gt; 2023-01-28T07:53:36+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7 !INFO | scylla[5380]: [shard 3] compaction_m &gt; anager - Starting off-strategy compaction for mview.users_by_last_name, 643 candidates were found &gt; &lt; 
t:2023-01-28 07:53:36,916 f:db_log_reader.py l:113 c:sdcm.db_log_reader p:DEBUG &gt; 2023-01-28T07:53:36+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7 !ERR | syslog-ng[4728]: Destination reliable &gt; queue full, dropping message; filename='/tmp/syslog-ng-00000.rqf', queue_len='129', mem_buf_size='100000', disk_buf_size='2000000', persist_name='afsocket_dd_qfile(stream,10.12.11.112:49153)' &gt; &lt; 
t:2023-01-28 07:53:36,916 f:db_log_reader.py l:113 c:sdcm.db_log_reader p:DEBUG &gt; 2023-01-28T07:53:36+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7 !ERR | syslog-ng[4728]: Destination reliable &gt; queue full, dropping message; filename='/tmp/syslog-ng-00000.rqf', queue_len='129', mem_buf_size='100000', disk_buf_size='2000000', persist_name='afsocket_dd_qfile(stream,10.12.11.112:49153)' &gt; &lt; 
t:2023-01-28 07:53:36,917 f:db_log_reader.py l:113 c:sdcm.db_log_reader p:DEBUG &gt; 2023-01-28T07:53:36+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7 !ERR | syslog-ng[4728]: Destination reliable &gt; queue full, dropping message; filename='/tmp/syslog-ng-00000.rqf', queue_len='130', mem_buf_size='100000', disk_buf_size='2000000', persist_name='afsocket_dd_qfile(stream,10.12.11.112:49153)' &gt; &lt; 
t:2023-01-28 07:53:36,917 f:db_log_reader.py l:113 c:sdcm.db_log_reader p:DEBUG &gt; 2023-01-28T07:53:36+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7 !ERR | syslog-ng[4728]: I/O error occurred wh &gt; ile writing; fd='35', error='Broken pipe (32)' &gt;
```

The bootstrap was interrapted, nothing run on the node after that.

The terminated node was not replaced finally, remained in the ring:
```
Datacenter: us-east
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.12.9.157   15.81 GB   256          ?       8285ac2e-341c-47a5-9178-1e6f028798b1  1c
DN  10.12.8.179  20.76 GB   256          ?       f1672d75-7088-4db0-8ad0-93c24b4a37d7  1c
UN  10.12.11.159  17.61 GB   256          ?       58f5e89d-3ec7-4dd8-a573-d42c72d09392  1c
UN  10.12.10.123  15.15 GB   256          ?       24e0082a-6279-4f69-8df1-f350e1ec47d7  1c
UN  10.12.8.241   17.84 GB   256          ?       2be4a7c5-02b5-4cd7-becf-49bd87923731  1c
```

3 hours later we started to get a lot of errors `Error applying view update to 10.12.8.179` (where ` 10.12.8.179` is terminated node:
```
2023-01-28 10:33:47.637 &lt;2023-01-28 10:33:47.000&gt;: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=4a7101eb-37b7-4af2-944a-5a1e89ce79b5: type=DATABASE_ERROR regex=(^ERROR|!\s*?ERR).*\[shard.*\] line_number=148359 node=longevity-tls-50gb-3d-master-db-node-91f4614d-4
2023-01-28T10:33:47+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-4      !ERR | scylla[746]:  [shard 10] view - Error applying view update to 10.12.8.179 (view: mview.users_by_last_name, base token: 6173207356020166693, view token: 4408317617141445361): seastar::broken_promise (broken promise)
```

There are **19961** those errors (`Error applying view update to 10.12.8.179`).

### View definition
There are 2 views in test. Its definition:
```
table_definition: |
  CREATE TABLE users (
    username text,
    first_name text,
    last_name text,
    password text,
    email text,
    last_access timeuuid,
    PRIMARY KEY(username)
  );
extra_definitions:
  - CREATE MATERIALIZED VIEW mview.users_by_first_name AS SELECT * FROM mview.users WHERE first_name IS NOT NULL and username IS NOT NULL PRIMARY KEY (first_name, username);
  - CREATE MATERIALIZED VIEW mview.users_by_last_name AS SELECT * FROM mview.users WHERE last_name IS NOT NULL and username IS NOT NULL PRIMARY KEY (last_name, username);

columnspec:
  - name: username
    size: uniform(10..30)
  - name: first_name
    size: fixed(16)
  - name: last_name
    size: uniform(1..32)
  - name: password
    size: fixed(80) # sha-512
  - name: email
    size: uniform(16..50)
```

Profile: 
https://github.com/scylladb/scylla-cluster-tests/blob/ce782d42d1aea97b0366fb51c80b70d521fdb749/data_dir/cs_mv_profile.yaml

This profile was run with command:
```
"cassandra-stress user profile=/tmp/cs_mv_profile.yaml ops'(insert=3,read1=1,read2=1,read3=1)' cl=QUORUM duration=4320m -mode cql3 native -rate threads=20"
```

## Impact

New node bootrap stuck. Terminated node was not replaced and stuck in the ring

## How frequently does it reproduce?

Maybe twice. In the previous run of this test the logs were not saved. But by consol output it's looks like the same issue.

**First failed run with similar issue** was run at `2023-01-20`:
https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-50gb-3days/177/
Scylla version: 5.3.0~dev-20230119.aab5954cfb65
Build id: 5363e4cdfb489648423400b04906f5a8aff16578

**Last succeeded run** was at `2023-01-13 `:
https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-50gb-3days/176/
Scylla version of succeeded run: 5.2.0~dev-20230112.95031074a504
Build id of succeeded run: 214a2060bce0394c5f505a48a88312cd2c9e0498

Commits between last succeeded and first failed:
[commits_between_aab5954cfb65_95031074a504.log](https://github.com/scylladb/scylladb/files/10556048/commits_between_aab5954cfb65_95031074a504.log)

## Installation details

Kernel Version: 5.15.0-1028-aws
Scylla version (or git commit hash): `5.3.0~dev-20230126.a0ff33e77774` with build-id `575b7a0891506801aeb1c05412a6eac011a0eda4`

Cluster size: 6 nodes (i4i.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-50gb-3d-master-db-node-91f4614d-7 (3.95.18.168 | 10.12.9.13) (shards: -1)
- longevity-tls-50gb-3d-master-db-node-91f4614d-6 (52.55.169.59 | 10.12.8.241) (shards: 14)
- longevity-tls-50gb-3d-master-db-node-91f4614d-5 (54.226.109.147 | 10.12.10.123) (shards: 14)
- longevity-tls-50gb-3d-master-db-node-91f4614d-4 (54.196.151.77 | 10.12.9.157) (shards: 14)
- longevity-tls-50gb-3d-master-db-node-91f4614d-3 (54.234.230.24 | 10.12.11.159) (shards: 14)
- longevity-tls-50gb-3d-master-db-node-91f4614d-2 (54.167.109.174 | 10.12.8.179) (shards: 14)
- longevity-tls-50gb-3d-master-db-node-91f4614d-1 (54.158.3.176 | 10.12.10.158) (shards: 14)

OS / Image: `ami-0a8fd7e11139b6ca1` (aws: us-east-1)

Test: `longevity-50gb-3days`
Test id: `91f4614d-70d0-4ae2-9c9b-b9aa15d603bd`
Test name: `scylla-master/longevity/longevity-50gb-3days`
Test config file(s):

- [longevity-50GB-3days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/ce782d42d1aea97b0366fb51c80b70d521fdb749/test-cases/longevity/longevity-50GB-3days-authorization-and-tls-ssl.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 91f4614d-70d0-4ae2-9c9b-b9aa15d603bd`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=91f4614d-70d0-4ae2-9c9b-b9aa15d603bd)
- Show all stored logs command: `$ hydra investigate show-logs 91f4614d-70d0-4ae2-9c9b-b9aa15d603bd`

## Logs:

- monitor-set: https://cloudius-jenkins-test.s3.amazonaws.com/91f4614d-70d0-4ae2-9c9b-b9aa15d603bd/20230129_164639/monitor-set-91f4614d.tar.gz
- db-cluster: https://cloudius-jenkins-test.s3.amazonaws.com/91f4614d-70d0-4ae2-9c9b-b9aa15d603bd/20230129_164639/longevity-tls-50gb-3d-master-db-cluster-91f4614d.tar.gz
- sct log: https://cloudius-jenkins-test.s3.amazonaws.com/91f4614d-70d0-4ae2-9c9b-b9aa15d603bd/20230129_164639/sct.log.tar.gz
- all events: https://cloudius-jenkins-test.s3.amazonaws.com/91f4614d-70d0-4ae2-9c9b-b9aa15d603bd/20230129_164639/events-91f4614d.log.tar.gz

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-50gb-3days/178/)
</details>

**Comment by juliayakovlev** [comment]
Issue https://github.com/scylladb/scylla-manager/issues/3294 happened before this one.

Same in the previous run of this job: 
https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-50gb-3days/177/

**Comment by avikivity** [comment]
It looks like a syslog-ng problem, not a scylladb problem.

**Comment by mykaul** [comment]
> It looks like a syslog-ng problem, not a scylladb problem.

Unless Scylla flooded it?

**Comment by fruch** [comment]
from POV of node-7:
```
2023-01-28T07:43:44+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7  !WARNING | scylla[5380]:  [shard  0] storage_service - The replace_address_first_boot=10.12.8.179 option is deprecated. Please use the replace_node_first_boot option
2023-01-28T07:43:44+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-7     !INFO | scylla[5380]:  [shard  0] storage_service - Gathering node replacement information for 00000000-0000-0000-0000-000000000000/10.12.8.179
```


and from the POV of node-6:
```
> grep '8bc343aa-bda1-4c81-86a0-d468c46849d8\|10.12.9.13' ~/system.log
2023-01-28T07:43:44+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6  !WARNING | scylla[745]:  [shard  6] raft_group_registry - Raft server id 8bc343aa-bda1-4c81-86a0-d468c46849d8 cannot be translated to an IP address.
2023-01-28T07:43:44+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6  !WARNING | scylla[745]:  [shard  6] raft_group_registry - Raft server id 8bc343aa-bda1-4c81-86a0-d468c46849d8 cannot be translated to an IP address.
2023-01-28T07:43:46+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] gossip - InetAddress 10.12.9.13 is now UP, status = UNKNOWN
2023-01-28T07:43:46+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] raft_group_registry - marking Raft server 8bc343aa-bda1-4c81-86a0-d468c46849d8 as alive for raft groups
2023-01-28T07:45:37+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] storage_service - replace[c3c41ffb-8c5a-4cbf-be3f-edb220ee1177]: Added replacing_node=10.12.9.13 to replace existing_node=10.12.8.179, coordinator=10.12.9.13
2023-01-28T07:45:37+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] token_metadata - Added node 10.12.9.13 as pending replacing endpoint which replaces existing node 10.12.8.179
```

I'm not sure if `cannot be translated to an IP address.` is o.k. in this sequence or not, or it's just wait for gossip communication with that node..

@bhalevy I don't know the sequence of replace has changed with the introduction of replace by host-id, but this failure in replacement operation right after this was merged
```
7510144fba59d9c1a60194fd6041b7a8d5aa479a Merge 'Add replace-node-first-boot option' from Benny Halevy
```

can you take a look maybe we are missing something here, even that the logs are a bit off, since they had to be extract manually.

**Comment by bhalevy** [comment]
There are several problems mentioned in this report, some of them could be cascading, but let's try to open separate issues for each since they would probably need to be fixed independently of each other.

1. replace failed.  What was the roit cause? Was this due to the syslog crash? Even if scylla triggered it by printing too much to the syslog it shouldn't crash. Was the failure caused by changes introduced in "replace-node-first-boot option", or is it just incidental?
2. view updates failed on the DN node, right? What should view updates so in this case? Cc @nyh / @kbr-scylla 
3. raft complaining. Is the host_id mentioned there is the one if the replacing node that failed to replace?  This seems yet another issue. Cc @kostja  @gleb-cloudius @kbr-scylla. This is a good test case too add to topology tests: do we test recovery from failed topology changes?

**Comment by fruch** [comment]
> There are several problems mentioned in this report, some of them could be cascading, but let's try to open separate issues for each since they would probably need to be fixed independently of each other.
> 
> 1. replace failed.  What was the roit cause? Was this due to the syslog crash? Even if scylla triggered it by printing too much to the syslog it shouldn't crash. Was the failure caused by changes introduced in "replace-node-first-boot option", or is it just incidental?

It's hard to tell, since the logging system broke, so we don't see a specific failure.

It's a case we run on a weekly basis, it worked cleanly one week, and now failed twice in a row in the same sequence (nemesis), in the first we didn't had any logs, the logs here are from the last failing run. (and going to run again this weekend, and I'm guess to fail again, we'll keep the nodes, so closer inspection can be done)

> 2. view updates failed on the DN node, right? What should view updates so in this case? Cc @nyh / @kbr-scylla 
> 3. raft complaining. Is the host_id mentioned there is the one if the replacing node that failed to replace? 

It's the host id of the new node failed to replace

> This seems yet another issue. Cc @kostja  @gleb-cloudius @kbr-scylla. This is a good test case too add to topology tests: do we test recovery from failed topology changes?

**Comment by gleb-cloudius** [comment]
> 
> and from the POV of node-6:
> 
> ```
> > grep '8bc343aa-bda1-4c81-86a0-d468c46849d8\|10.12.9.13' ~/system.log
> 2023-01-28T07:43:44+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6  !WARNING | scylla[745]:  [shard  6] raft_group_registry - Raft server id 8bc343aa-bda1-4c81-86a0-d468c46849d8 cannot be translated to an IP address.
> 2023-01-28T07:43:44+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6  !WARNING | scylla[745]:  [shard  6] raft_group_registry - Raft server id 8bc343aa-bda1-4c81-86a0-d468c46849d8 cannot be translated to an IP address.
> 2023-01-28T07:43:46+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] gossip - InetAddress 10.12.9.13 is now UP, status = UNKNOWN
> 2023-01-28T07:43:46+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] raft_group_registry - marking Raft server 8bc343aa-bda1-4c81-86a0-d468c46849d8 as alive for raft groups
> 2023-01-28T07:45:37+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] storage_service - replace[c3c41ffb-8c5a-4cbf-be3f-edb220ee1177]: Added replacing_node=10.12.9.13 to replace existing_node=10.12.8.179, coordinator=10.12.9.13
> 2023-01-28T07:45:37+00:00 longevity-tls-50gb-3d-master-db-node-91f4614d-6     !INFO | scylla[745]:  [shard  0] token_metadata - Added node 10.12.9.13 as pending replacing endpoint which replaces existing node 10.12.8.179
> ```
> 
> I'm not sure if `cannot be translated to an IP address.` is o.k. in this sequence or not, or it's just wait for gossip communication with that node..
> 

If they disappeared after `8bc343aa-bda1-4c81-86a0-d468c46849d8` was reported as up it is not a problem.

**Comment by nyh** [comment]
> There are several problems mentioned in this report, some of them could be cascading, but let's try to open separate issues for each since they would probably need to be fixed independently of each other.
> ...
>     2. view updates failed on the DN node, right? What should view updates so in this case? Cc @nyh / @kbr-scylla

This is a good question. I see this was already reported in the past in an issue I wasn't aware of (https://github.com/scylladb/scylladb/issues/12321). Clearly, view update failure (the base replica sending an update to a view replica) is a normal event that happens whenever nodes go down, and we even have view hints to solve this case (in some sense), so spamming the log in this case isn't a great idea. But maybe we did want to show *something* in the log? I don't remember. I'm not sure that this "broken promise" spam isn't a new regression, I'll have to take a look at it. But I don't know if the log spam is an important part of this issue, really?

**Comment by fruch** [comment]
> > There are several problems mentioned in this report, some of them could be cascading, but let's try to open separate issues for each since they would probably need to be fixed independently of each other.
> > ...
> >     2. view updates failed on the DN node, right? What should view updates so in this case? Cc @nyh / @kbr-scylla
> 
> This is a good question. I see this was already reported in the past in an issue I wasn't aware of (https://github.com/scylladb/scylladb/issues/12321). Clearly, view update failure (the base replica sending an update to a view replica) is a normal event that happens whenever nodes go down, and we even have view hints to solve this case (in some sense), so spamming the log in this case isn't a great idea. But maybe we did want to show *something* in the log? I don't remember. I'm not sure that this "broken promise" spam isn't a new regression, I'll have to take a look at it. But I don't know if the log spam is an important part of this issue, really?

The log spam is important in this case, since it's breaking SCT log collection system and inflating the sizes of logs to the degree we fail to collect them

And why in turn make the investigation of the actual issue much much harder.

Yes it's not the root cause, and we can point to a specific time those started happening

**Comment by DoronArazii** [comment]
@asias @raphaelsc can you please check

**Comment by nyh** [comment]
> The log spam is important in this case, since it's breaking SCT log collection system and inflating the sizes of logs to the degree we fail to collect them

Opened https://github.com/scylladb/scylladb/issues/12711.

**Comment by raphaelsc** [comment]
> > It looks like a syslog-ng problem, not a scylladb problem.
> 
> Unless Scylla flooded it?

reshape compaction logging can be long, as it can have RF*nr_tokens input sstables at once. 

A typical sstable in the log goes like this: "/var/lib/data/ks/scylla-e6a3044a-1108-46cf-bacb-4ab62559fbbe/mc-5-big-Data.db:level=0" ~90 bytes.

With RF=3, nr_tokens=256, we have 768 input files = 67k of logging data per shard

With 14 shards, 945k of logging data, sent roughly at the same time.

Judging from this message "queue full, dropping message; filename='/tmp/syslog-ng-00000.rqf', queue_len='129', mem_buf_size='100000', disk_buf_size='2000000',", that amount of data seems more than enough to put stress on syslog-ng queues. 

we can compress the compaction logging, printing directory only once, or just a hard trimming to N bytes which is not optimal as it loses data.

**Comment by fruch** [comment]
It happened again this weekend

the machines are up, and you can connection with `ssh:
```
   longevity-tls-50gb-3d-master-db-node-beb6cd3d-4 - 54.204.239.68 10.12.9.230 - us-east-1
   longevity-tls-50gb-3d-master-db-node-beb6cd3d-1 - 18.212.99.214 10.12.10.148 - us-east-1
   longevity-tls-50gb-3d-master-db-node-beb6cd3d-2 - 54.234.68.218 10.12.10.59 - us-east-1
   longevity-tls-50gb-3d-master-loader-node-beb6cd3d-1 - 54.147.14.71 10.12.11.55 - us-east-1
   longevity-tls-50gb-3d-master-loader-node-beb6cd3d-2 - 54.242.229.226 10.12.9.150 - us-east-1
   longevity-tls-50gb-3d-master-loader-node-beb6cd3d-3 - 3.82.48.204 10.12.8.120 - us-east-1
   longevity-tls-50gb-3d-master-monitor-node-beb6cd3d-1 - 54.234.168.88 10.12.9.122 - us-east-1
   longevity-tls-50gb-3d-master-db-node-beb6cd3d-7 - 54.146.223.37 10.12.10.132 - us-east-1
   longevity-tls-50gb-3d-master-db-node-beb6cd3d-8 - 107.22.66.226 10.12.10.155 - us-east-1
   longevity-tls-50gb-3d-master-db-node-beb6cd3d-10 - 54.90.73.71 10.12.11.21 - us-east-1
```

 longevity-tls-50gb-3d-master-db-node-beb6cd3d-10 - 54.90.73.71 10.12.11.21  is replacing 36cc408f-2e23-45f7-8952-66e27a36e4b8/10.12.10.245

the old node still seen in `nodetool status`, while the new node doesn't appear:
```
Datacenter: us-east
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.12.10.59   16.86 GB   256          ?       574195b0-d3f0-472e-8779-67b849d0238d  1c
UN  10.12.10.155  17.77 GB   256          ?       9a8b0c3e-6960-4dc8-a87f-a17a003ab698  1c
UN  10.12.10.148  17.38 GB   256          ?       bddd08de-6277-4912-8cf0-4065d55716bf  1c
UN  10.12.10.132  17.68 GB   256          ?       c70fb6fc-4861-4ccd-8e86-511c090c4e41  1c
DN  10.12.10.245  18.57 GB   256          ?       36cc408f-2e23-45f7-8952-66e27a36e4b8  1c
UN  10.12.9.230   17.27 GB   256          ?       bf07d6fa-f4ac-46d6-b012-341df62d9fdb  1c
```

on the gossip info (done on 10.12.10.155), we can see one node in shutdown, and one node doesn't have status.
```
/10.12.10.245
  generation:0
  heartbeat:0
  X6:14
  HOST_ID:36cc408f-2e23-45f7-8952-66e27a36e4b8
  SCHEMA:323b145c-48f6-3042-a406-84ba51aa190a
  RPC_ADDRESS:10.12.10.245
  NET_VERSION:0
  X4:0
  RELEASE_VERSION:3.0.8
  X9:org.apache.cassandra.locator.Ec2Snitch
  DC:us-east
  RACK:1c
  X2:system_traces.events:0.000000;mview.users_by_last_name:0.994227;mview.users:0.999913;mview.users_by_first_name:0.993767;system_traces.node_slow_log_time_idx:0.000000;system_traces.sessions_time_idx:0.000000;system_auth.role_attributes:0.000000;system_auth.role_members:0.000000;ks_truncate_large_partition.test_table:0.000000;keyspace1.standard1:0.713575;system_traces.sessions:0.000000;ks_truncate_large_partition.test_counters:0.000000;system_auth.role_permissions:0.000000;system_traces.node_slow_log:0.000000;system_auth.roles:0.000000;system_distributed.cdc_generation_timestamps:0.000000;system_distributed.view_build_status:0.000000;system_distributed_everywhere.cdc_generation_descriptions_v2:0.000000;system_distributed.service_levels:1.000000;ks_truncate.standard1:0.000000;system_distributed.cdc_streams_descriptions_v2:0.000000;
  X3:3
  STATUS:shutdown,true
  X7:12
  X1:AGGREGATE_STORAGE_OPTIONS,ALTERNATOR_STREAMS,ALTERNATOR_TTL,CDC,CDC_GENERATIONS_V2,COLLECTION_INDEXING,COMPUTED_COLUMNS,CORRECT_COUNTER_ORDER,CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX,CORRECT_NON_COMPOUND_RANGE_TOMBSTONES,CORRECT_STATIC_COMPACT_IN_MC,COUNTERS,DIGEST_FOR_NULL_VALUES,DIGEST_INSENSITIVE_TO_EXPIRY,DIGEST_MULTIPARTITION_READ,EMPTY_REPLICA_PAGES,HINTED_HANDOFF_SEPARATE_CONNECTION,INDEXES,KEYSPACE_STORAGE_OPTIONS,LARGE_COLLECTION_DETECTION,LARGE_PARTITIONS,LA_SSTABLE_FORMAT,LWT,MATERIALIZED_VIEWS,MC_SSTABLE_FORMAT,MD_SSTABLE_FORMAT,ME_SSTABLE_FORMAT,NONFROZEN_UDTS,PARALLELIZED_AGGREGATION,PER_TABLE_CACHING,PER_TABLE_PARTITIONERS,RANGE_SCAN_DATA_VARIANT,RANGE_TOMBSTONES,ROLES,ROW_LEVEL_REPAIR,SCHEMA_COMMITLOG,SCHEMA_TABLES_V3,SECONDARY_INDEXES_ON_STATIC_COLUMNS,SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT,STREAM_WITH_RPC_STREAM,SUPPORTS_RAFT_CLUSTER_MANAGEMENT,TOMBSTONE_GC_OPTIONS,TRUNCATION_TABLE,TYPED_ERRORS_IN_READ_RPC,UDA,UDA_NATIVE_PARALLELIZED_AGGREGATION,UNBOUNDED_RANGE_TOMBSTONES,VIEW_VIRTUAL_COLUMNS,WRITE_FAILURE_REPLY,XXHASH
  X8:v2;1676141261313;ea0e8387-be58-4efe-b7bf-2d867103edab
  LOAD:19942492160
  X5:1544:882691276:1676238205589

...
/10.12.11.21
  generation:0
  heartbeat:0
  X6:14
  HOST_ID:2807b6d9-699a-43c0-a7aa-343ed6dbdac5
  SCHEMA:323b145c-48f6-3042-a406-84ba51aa190a
  RPC_ADDRESS:10.12.11.21
  NET_VERSION:0
  X4:0
  RELEASE_VERSION:3.0.8
  X9:org.apache.cassandra.locator.Ec2Snitch
  DC:us-east
  RACK:1c
  X2:
  X3:3
  X7:12
  X1:AGGREGATE_STORAGE_OPTIONS,ALTERNATOR_STREAMS,ALTERNATOR_TTL,CDC,CDC_GENERATIONS_V2,COLLECTION_INDEXING,COMPUTED_COLUMNS,CORRECT_COUNTER_ORDER,CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX,CORRECT_NON_COMPOUND_RANGE_TOMBSTONES,CORRECT_STATIC_COMPACT_IN_MC,COUNTERS,DIGEST_FOR_NULL_VALUES,DIGEST_INSENSITIVE_TO_EXPIRY,DIGEST_MULTIPARTITION_READ,EMPTY_REPLICA_PAGES,HINTED_HANDOFF_SEPARATE_CONNECTION,INDEXES,KEYSPACE_STORAGE_OPTIONS,LARGE_COLLECTION_DETECTION,LARGE_PARTITIONS,LA_SSTABLE_FORMAT,LWT,MATERIALIZED_VIEWS,MC_SSTABLE_FORMAT,MD_SSTABLE_FORMAT,ME_SSTABLE_FORMAT,NONFROZEN_UDTS,PARALLELIZED_AGGREGATION,PER_TABLE_CACHING,PER_TABLE_PARTITIONERS,RANGE_SCAN_DATA_VARIANT,RANGE_TOMBSTONES,ROLES,ROW_LEVEL_REPAIR,SCHEMA_COMMITLOG,SCHEMA_TABLES_V3,SECONDARY_INDEXES_ON_STATIC_COLUMNS,SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT,STREAM_WITH_RPC_STREAM,SUPPORTS_RAFT_CLUSTER_MANAGEMENT,TOMBSTONE_GC_OPTIONS,TRUNCATION_TABLE,TYPED_ERRORS_IN_READ_RPC,UDA,UDA_NATIVE_PARALLELIZED_AGGREGATION,UNBOUNDED_RANGE_TOMBSTONES,VIEW_VIRTUAL_COLUMNS,WRITE_FAILURE_REPLY,XXHASH
  X5:
```

**Comment by fruch** [comment]
@asias @raphaelsc, can you guys take a closer look here ? (I don't have the logs, since the of all the issue mentioned here, that breaking our logs system), but the nodes are up, and in a unfinished replacement state.

**Comment by fruch** [comment]
Took the cluster down, after I collect the logs manually

## Logs
* **db-cluster** - https://cloudius-jenkins-test.s3.amazonaws.com/beb6cd3d-dc11-4efc-b5ee-9dfb1ab42c9d/20230214_224220/db-cluster-beb6cd3d.tar.gz
* **loader-set** - https://cloudius-jenkins-test.s3.amazonaws.com/beb6cd3d-dc11-4efc-b5ee-9dfb1ab42c9d/20230214_224220/loader-set-beb6cd3d.tar.gz
* **monitor-set** - https://cloudius-jenkins-test.s3.amazonaws.com/beb6cd3d-dc11-4efc-b5ee-9dfb1ab42c9d/20230214_224220/monitor-set-beb6cd3d.tar.gz

**Comment by Michal-Leszczynski** [comment]
The Scylla Manager [issue](https://github.com/scylladb/scylla-manager/issues/3294) mentioned earlier seems to be caused because SM awaits Scylla to finish its repair job which is not happening. Scylla keeps on returning "RUNNING" status for this job, so SM keeps on waiting for its end, but it does not happen for many hours.
Here is fragment of logs that indicates that (in original logs there are about 4k (over 11h) queries for each non ending repair job):

```
2023-02-12T20:11:56+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:11:56.134Z","N":"http","M":"GET /storage_service/repair_status?id=377&timeout=10","from":"10.4.3.222:43000","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:11:56+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:11:56.134Z","N":"http","M":"GET /storage_service/repair_status?id=378&timeout=10","from":"10.4.3.222:36610","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:11:56+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:11:56.491Z","N":"http","M":"GET /storage_service/repair_status?id=379&timeout=10","from":"10.4.3.222:60542","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:06+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:06.135Z","N":"http","M":"GET /storage_service/repair_status?id=377&timeout=10","from":"10.4.3.222:43000","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:06+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:06.135Z","N":"http","M":"GET /storage_service/repair_status?id=378&timeout=10","from":"10.4.3.222:36610","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:06+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:06.492Z","N":"http","M":"GET /storage_service/repair_status?id=379&timeout=10","from":"10.4.3.222:60542","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:16+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:16.135Z","N":"http","M":"GET /storage_service/repair_status?id=377&timeout=10","from":"10.4.3.222:43000","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:16+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:16.136Z","N":"http","M":"GET /storage_service/repair_status?id=378&timeout=10","from":"10.4.3.222:36610","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:16+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:16.493Z","N":"http","M":"GET /storage_service/repair_status?id=379&timeout=10","from":"10.4.3.222:60542","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:26+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:26.136Z","N":"http","M":"GET /storage_service/repair_status?id=378&timeout=10","from":"10.4.3.222:44370","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:26+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:26.136Z","N":"http","M":"GET /storage_service/repair_status?id=377&timeout=10","from":"10.4.3.222:43000","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:26+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:26.494Z","N":"http","M":"GET /storage_service/repair_status?id=379&timeout=10","from":"10.4.3.222:60542","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:36+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:36.137Z","N":"http","M":"GET /storage_service/repair_status?id=378&timeout=10","from":"10.4.3.222:44370","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:36+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:36.137Z","N":"http","M":"GET /storage_service/repair_status?id=377&timeout=10","from":"10.4.3.222:43000","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:36+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:36.495Z","N":"http","M":"GET /storage_service/repair_status?id=379&timeout=10","from":"10.4.3.222:60542","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:46+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:46.138Z","N":"http","M":"GET /storage_service/repair_status?id=377&timeout=10","from":"10.4.3.222:43000","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:46+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:46.138Z","N":"http","M":"GET /storage_service/repair_status?id=378&timeout=10","from":"10.4.3.222:44370","status":200,"bytes":9,"duration":"10000ms"}
2023-02-12T20:12:46+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T20:12:46.495Z","N":"http","M":"GET /storage_service/repair_status?id=379&timeout=10","from":"10.4.3.222:60542","status":200,"bytes":9,"duration":"10000ms"}
```
It's important to note that 12 bytes in response indicates "SUCCESSFUL" status and is usually returned before 10 seconds passed. On the other hand, "RUNNING" status takes only 9 bytes and is returned after exactly 10 seconds.
Here is an example of "SUCCESSFUL" job status:

```
2023-02-12T08:41:17+00:00 longevity-tls-50gb-3d-5-2-db-node-0530dbf7-16     !INFO | scylla-manager-agent[828]: {"L":"INFO","T":"2023-02-12T08:41:17.893Z","N":"http","M":"GET /storage_service/repair_status?id=373&timeout=10","from":"10.4.3.222:35900","status":200,"bytes":12,"duration":"1364ms"}
```
I wanted to paste it here because it might be helpful for someone assigned to this task on Scylla core side. Should I make a separate issue for this?

Logs above were taken from 16'th node that can be found here:  https://cloudius-jenkins-test.s3.amazonaws.com/0530dbf7-d9a0-429a-8d7b-47e16846ade2/20230212_201319/db-cluster-0530dbf7.tar.gz

**Comment by roydahan** [comment]
@asias can you please take a look at @Michal-Leszczynski comment above and tell us if it requires a new issue separated from this one?

**Comment by roydahan** [comment]
Someone should tell why it's not blocker for 5.2

**Comment by DoronArazii** [comment]
If it's a blocker for 5.2, the priority should increase to P1 by now.

@roydahan AFAIU it's happening on master, which is 5.3 (Labeled "Master/Triage") - so irrelevant for 5.2 ?

**Comment by mykaul** [comment]
> If it's a blocker for 5.2, the priority should increase to P1 by now.
> 
> @roydahan AFAIU it's happening on master, which is 5.3 (Labeled "Master/Triage") - so irrelevant for 5.2 ?

It was also reported on 5.2, so clearly relevant.

**Comment by mykaul** [comment]
I don't think anyone looked at why repair from Scylla side did not seem to end.

**Comment by DoronArazii** [comment]
@eliransin / @asias can you please assist

**Comment by kostja** [comment]
@DoronArazii this is raft domain. The code works according to the original design. We could automatically remove down nodes without tokens from group0 after an interval, if that's what you want, it's an additional feature we need to implement.

**Comment by roydahan** [comment]
@Michal-Leszczynski, per your comment [here](https://github.com/scylladb/scylladb/issues/12693#issuecomment-1433244239) I suggest you'll open a new bug in scylla for the issue that may be the cause for https://github.com/scylladb/scylla-manager/issues/3294.

One of them is a blocker for the release.

**Comment by roydahan** [comment]
> @DoronArazii this is raft domain. The code works according to the original design. We could automatically remove down nodes without tokens from group0 after an interval, if that's what you want, it's an additional feature we need to implement.

@kostja is there a way to at least rate limit these amount of errors?
And possibly reduce their severity.

**Comment by fruch** [comment]
> @DoronArazii this is raft domain. The code works according to the original design. We could automatically remove down nodes without tokens from group0 after an interval, if that's what you want, it's an additional feature we need to implement.

@kostja can you elaborate on how it affects this replacement case ? with gossip it would be working ?

**Comment by kostja** [comment]
Folks I am terribly sorry I commented on a wrong issue.

**Comment by kostja** [comment]
> > @DoronArazii this is raft domain. The code works according to the original design. We could automatically remove down nodes without tokens from group0 after an interval, if that's what you want, it's an additional feature we need to implement.
> 
> @kostja can you elaborate on how it affects this replacement case ? with gossip it would be working ?

I mixed this one and https://github.com/scylladb/scylladb/issues/13120 :(

**Comment by DoronArazii** [comment]
@asias / @raphaelsc can you please reffer to @Michal-Leszczynski [comment](https://github.com/scylladb/scylladb/issues/12693#issuecomment-1433244239)

**Comment by roydahan** [comment]
There are 3 problems here (and I think we should split them into 3 different issues):
1. The original reported bootstrap failure - **High priority**.
2. The log spamming with huge amount of "Error applying view update", the amount is so big SCT is OOM'ing on processing this - **High Priority**.
3. Manager Repair stuck, not getting response from scylla, suspected by the same root cause, but not proved yet - **High Priority**.

@juliayakovlev please split the second one from this issue (keep the reference).
@Michal-Leszczynski please open a new scylla core issue for https://github.com/scylladb/scylla-manager/issues/3294 in order for someone to investigate the root cause (keep the reference for this issue).

**Comment by juliayakovlev** [comment]
> There are 3 problems here (and I think we should split them into 3 different issues):
> 
> 1. The original reported bootstrap failure - **High priority**.
> 2. The log spamming with huge amount of "Error applying view update", the amount is so big SCT is OOM'ing on processing this - **High Priority**.

This problem looks similar to https://github.com/scylladb/scylladb/issues/12321

> 3. Manager Repair stuck, not getting response from scylla, suspected by the same root cause, but not proved yet - **High Priority**.
> 
> @juliayakovlev please split the second one from this issue (keep the reference). @Michal-Leszczynski please open a new scylla core issue for [scylladb/scylla-manager#3294](https://github.com/scylladb/scylla-manager/issues/3294) in order for someone to investigate the root cause (keep the reference for this issue).

**Comment by DoronArazii** [comment]
@raphaelsc can you please help with this / share ideas?
This issue is a blocker for 5.2

/Cc @bhalevy

**Comment by bhalevy** [comment]
@denesb, @raphaelsc is away. Can you please look into this?
It may be related to unlimited concurrency of view update reads, but we also need to explain the apparent regression.
Maybe it's combined with another reason that triggered it in 5.2

**Comment by eliransin** [comment]
@asias , @denesb this is a high priority blocker for 5.2, can you please look at this and try come up with the root cause,
or plan next steps?

**Comment by denesb** [comment]
I will have a look.

**Comment by bhalevy** [comment]
Let's just define the main issue here as "replace node failing".
The "bootstrap" here is the one that happens on the replacing node as part of the replace operation according to @roydahan

**Comment by bhalevy** [comment]
"Error applying view update (seastar::broken_promise)" is tracked by:
#12321 and #12711
I'm not sure if and how it's related to replace failing

**Comment by bhalevy** [comment]
https://github.com/scylladb/scylladb/pull/13175 adds rate limiting to the above error message

**Comment by kbr-scylla** [comment]
> The replace uses RBNO, which was recently enabled. 

Note that replace was already using RBNO by default for some time.
What changed recently is enabling RBNO by default for other operations: bootstrap, removenode, decommission.

**Comment by denesb** [comment]
Like I said in my comment above, a coredump would be of great help in getting more insight into what is happening.
In the meanwhile I had a wild theory which I want to look into, but didn't get the chance yet as I was busy with other stuff.

**Comment by DoronArazii** [comment]
@juliayakovlev ^^

**Comment by denesb** [comment]
@asias I'm looking at https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c. I see no place in the code where the permit is relinquished. Can there be a situation where all permits are taken but the repair want to read from a new shard and need to obtain a permit? In this case there would be a deadlock as all permits are gone but the repair as a whole is waiting on a new one.

**Comment by roydahan** [comment]
@juliayakovlev can you please try to reproduce the original issue and possibly remove the timeout so we can provide a live cluster where it stuck? (I don't trust our SIGABRT to produce a coredump).

@denesb do you think that the stuck repair of the manager issue that is mentioned above, really related to this issue?
(If so, it'll be much easier to reproduce).

**Comment by asias** [comment]
> @asias I'm looking at [c6087cf](https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c). I see no place in the code where the permit is relinquished. Can there be a situation where all permits are taken but the repair want to read from a new shard and need to obtain a permit? In this case there would be a deadlock as all permits are gone but the repair as a whole is waiting on a new one.

The permit will be relinquished when the std::optional<repair_reader_permit_meta> _reader_permit in repair_meta is destroyed.  Each repair_meta will take all the permits it needs in one shot.

**Comment by asias** [comment]
All nodes have 13 shards, so the cluster was executing the easy path of commit [c6087cf](https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c), which takes only local lock.

**Comment by asias** [comment]
At 2023.1.27 07:15, node 4 with ip 9.157 started to have  view_update_generator_pending_registrations == 10 on shard 9. It never decreased since then. This was going to make any repair which triggers view update stuck forever.  Replace operation on node 7 with ip 9.13 was not supposed to be stuck by the stuck view_update_generator_pending_registrations though. 

```
    _metrics.add_group("view_update_generator", {
        sm::make_gauge("pending_registrations", sm::description("Number of tasks waiting to register staging sstables"),
                [this] { return _registration_sem.waiters(); }), 
```

![image](https://user-images.githubusercontent.com/320076/226265422-cfd86755-28f6-4238-bf6c-f7a7100a992d.png)

**Comment by denesb** [comment]
> > @asias I'm looking at [c6087cf](https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c). I see no place in the code where the permit is relinquished. Can there be a situation where all permits are taken but the repair want to read from a new shard and need to obtain a permit? In this case there would be a deadlock as all permits are gone but the repair as a whole is waiting on a new one.
> 
> The permit will be relinquished when the std::optional<repair_reader_permit_meta> _reader_permit in repair_meta is destroyed. Each repair_meta will take all the permits it needs in one shot.

But isn't repair meta alive for the entire duration of the repair? So in practice the lock is held for the entire repair too?

**Comment by asias** [comment]
> > > @asias I'm looking at [c6087cf](https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c). I see no place in the code where the permit is relinquished. Can there be a situation where all permits are taken but the repair want to read from a new shard and need to obtain a permit? In this case there would be a deadlock as all permits are gone but the repair as a whole is waiting on a new one.
> > 
> > 
> > The permit will be relinquished when the std::optional<repair_reader_permit_meta> _reader_permit in repair_meta is destroyed. Each repair_meta will take all the permits it needs in one shot.
> 
> But isn't repair meta alive for the entire duration of the repair? So in practice the lock is held for the entire repair too?

Yes, this is how is supposed to work to avoid creating more readers which could cause reader evictions. 

However, I think I found a possible race condition.

Consider:

Node n1 and n2 in the cluster, 1 shard per node, rf = 2, each shard has 1
available unit for the reader lock

- n1 starts repair r1
- r1-n1 (instance of r1 on node1) takes the reader lock on node1
- n2 starts repair r2 
- r2-n2 (instance of r2 on node2) takes the reader lock on node2
- r1-n2 will fail to take the reader lock on node2
- r2-n1 will fail to take the reader lock on node1

As a result, r1 and r2 could not make progress and deadlock happens.

The complexity comes from the fact that a repair job needs lock on more than one node. It is not
guaranteed that all the participant nodes could take the lock in one short.

**Comment by denesb** [comment]
Right, I forgot to consider the multi-node aspect.
This multi-node deadlock happened in the past, this is why we had to make repair readers evictable if you recall. We also considered a multi-node lock back then, but decided against it due to its complexity.

**Comment by denesb** [comment]
I think we will have to revert https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c and come up with a different solution for dealing with reader trashing. A whole cluster lock is not something we can come up in a quick-fix time-table.

**Comment by asias** [comment]
> I think we will have to revert [c6087cf](https://github.com/scylladb/scylladb/commit/c6087cf3a05d2022555189b4d240dbc836a9c72c) and come up with a different solution for dealing with reader trashing. A whole cluster lock is not something we can come up in a quick-fix time-table.

Yes, I think the same here.

**Comment by denesb** [comment]
I opened a PR with the revert: https://github.com/scylladb/scylladb/pull/13266

**Comment by denesb** [comment]
Backported to 5.2, other releases are not affected.

**Comment by asias** [comment]
> Backported to 5.2, other releases are not affected.

We need the backport for 2022.2 branch. scylla-2022.2.2 is affected.

**Comment by denesb** [comment]
> > Backported to 5.2, other releases are not affected.
> 
> @denesb We need the backport for 2022.2 branch. scylla-2022.2.2 is affected.

Right, I checked 5.1 which didn't have this, but we apparently backported it to 2022.2. I backported the revert to 2022.2 now.

**Comment by mykaul** [comment]
@DoronArazii - https://github.com/scylladb/scylladb/issues/12693#issuecomment-1502886122 seems to need an Enterprise issue tracking it (for 2022.2.5 I reckon)

**Comment by roydahan** [comment]
The stuck mgmt repair issue (https://github.com/scylladb/scylla-manager/issues/3294) still reproduce after this revert, so it's probably not related to it.
@Michal-Leszczynski / @asias if we blame scylla core for that issue, we need an issue number and a fix ASAP.
